KArray/KArrayCompile.lean

```diff
@@ -1,21 +1,99 @@
+  | some n => toString n
+  | none => toString e
+
+/-
+# TODO: make sure compiled functions only depend on either:
+* functions supported out-of-the-box like add, mul, sqrt, cos etc or
+* other compiled functions
+-/
+partial def toCCode (compilationUnits : List CompilationUnit)
+(e : Expr) : MetaM String := do
+  match e with
+  | Expr.fvar _ _=> toString e
+  | Expr.app f x _ => do
+    -- Is `f` reflected?
+    let s ← synthInstance? (← mkAppM `Reflected #[f])
+    match s with
+    | some _ => ename f ++ "(" ++ (← toCCode compilationUnits x) ++ ")"
+    | none => do
+      (← toCCode compilationUnits f) ++ "(" ++ (← toCCode compilationUnits x) ++ ")"
+  | _ => throwError "Invalid Expression"
+
+partial def metaCompile (compilationUnits : List CompilationUnit)
+(declName : Name) : MetaM String := do
+  let metaExpr ← whnf $ mkConst declName
+  match metaExpr with
+  | Expr.lam .. => lambdaTelescope metaExpr fun xs b => do (toCCode compilationUnits b)
+  | _ => throwError "Function expected!"
+
+def collectCompilationUnits (filePath : FilePath) : IO (List CompilationUnit) := do
+  let input ← IO.FS.readFile filePath
+  let (env, ok) ← Lean.Elab.runFrontend input Options.empty filePath.toString `main
+  if ok then
+    let mut res : List CompilationUnit ← []
+    let externMap ← externAttr.ext.getState env
+    for declName in kArrayCompileAttr.ext.getState env do
+      match externMap.find? declName with
+      | none =>
+        IO.println $ s!"Warning ({filePath}):\n\t`{declName}` isn't marked with " ++
+          "`extern`. Skipping KArray compilation."
+      | some data =>
+        let mut hasProperStandardEntry ← false
+        for entry in data.entries do
+          match entry with
+          | ExternEntry.standard name targetName =>
+            let nameString ← name.toString
+            if nameString = "all" ∨ nameString = "cpp" then -- this condition suffices
+              res ← res.concat ⟨filePath, env, declName, targetName⟩
+              hasProperStandardEntry ← true
+              break
+          | _ => continue -- non-standard entries are ignored
+        if ¬hasProperStandardEntry then
+          panic! s!"\nError ({filePath}):\n\t`kcompile` tag requires `{declName}` to " ++
+            "be marked with `extern <targetName>`."
+    res
+  else
+    panic! s!"\nLean's Frontend failed to run `{filePath}`."
 
-constant semicolon : String := ";"
+-- TODO: check if `targetName` is a valid function name for C code
+def isValidCFunctionName (targetName : String) : Bool := true
 
-constant newLine : String := "\n"
+def countOcurrences (l : List String) (n : String) : IO Nat := do
+  let mut c : Nat := 0
+  for ni in l do
+    if ni = n then
+      c := c + 1
+  c
 
-constant semicolonNewLine : String := semicolon ++ newLine
+def validateCompilationUnits (compilationUnits : List CompilationUnit) : IO PUnit := do
+  let targetNames ← compilationUnits.map λ c => c.targetName
+  for compilationUnit in compilationUnits do
```

Comment:
This is O(N^2) algorithm, probably not an issue right now as we are not processing thousands of functions. If it becomes a problem, maybe storing CompilationUnit in a hash map or what ever Lean uses to implement 'a set'.